### ISSUE TITLE
automate TYPE_TO_STRING

### ISSUE THREAD

**Issue by Trass3r** [onqtam/doctest#394]
Just read the docs on type parameterized tests.
With recent compilers it's [possible](https://gist.github.com/Trass3r/6fae8f543863a8853226d8c086efe7b1#file-enumflags-h-L17) to implement TYPE_TO_STRING by parsing __PRETTY_FUNCTION__. So perhaps that could be used to make custom TYPE_TO_STRINGs obsolete in the future.

**Comment by martinus** [comment]
It would be possible with `typeid` and `abi::__cxa_demangle`, see this: https://stackoverflow.com/a/4541470/48181

Note that this would resolve `typedef`s, so the output is not necessarily  the same as the input.

**Comment by Trass3r** [comment]
That's expensive, platform-specific and doesn't work with `-fno-rtti`.
But since we're using macros anyway one could also iterate over the types given to `TEST_CASE_TEMPLATE` and call `TYPE_TO_STRING` for each of them. Only problem is with the current approach you get redefinition errors for `type_to_string`.

**Comment by Trass3r** [comment]
That's expensive, platform-specific and doesn't work with `-fno-rtti`.

But since we're using macros anyway one could also iterate over the types given to `TEST_CASE_TEMPLATE` and call `TYPE_TO_STRING` for each of them. Only problem is you'd get redefinition errors for `type_to_string` if types occur more than once in the file.

Proof of concept: https://godbolt.org/z/s57rYa
```cpp
#define MY_TYPE_TO_STRING(...) \
    namespace doctest { namespace detail {                                                         \
            DOCTEST_TYPE_TO_STRING_IMPL(__VA_ARGS__)                                               \
        }                                                                                          \
    }

#undef DOCTEST_TEST_CASE_TEMPLATE
#define DOCTEST_TEST_CASE_TEMPLATE(dec, T, ...)                                                    \
    MAP(MY_TYPE_TO_STRING, __VA_ARGS__)                                                            \
    DOCTEST_TEST_CASE_TEMPLATE_IMPL(dec, T, DOCTEST_ANONYMOUS(_DOCTEST_ANON_TMP_), __VA_ARGS__)
```

**Comment by onqtam** [comment]
My original plan was indeed to use `abi::__cxa_demangle` but I avoided it because it required additional header includes in the interface part of doctest which is a no-no, but now that we'll be adding [extension headers](https://github.com/onqtam/doctest/issues/405) I guess this could be done. I think GoogleTest also uses `abi::__cxa_demangle` - that's what I looked at when implementing the template test cases.
Will be happy for a PR :D

**Comment by Trass3r** [comment]
Well the problem with `__PRETTY_FUNCTION__` is that gcc fixed its use in constexpr functions quite late in v9 and decided not to backport.
Your best bet would be https://github.com/onqtam/doctest/issues/394#issuecomment-665105469 but I don't know how to solve the fundamental redefinition problem.

**Comment by Saalvage** [comment]
@Trass3r I'm currently in the process of fixing up all of doctest's stringification stuff, which includes a better approach to stringifying types. At first I implemented it via `typeid` and `abi::__cxa_demangle`. I know this is quite old, but do you still think that the approach utilizing `__PRETTY_FUNCTION__` (and `__FUNCSIG__` on MSVC) is superior?

`constexpr` being broken shouldn't really be an issue tbh, and I don't really see the issue with redefinitions, if the implementation just remains a single templated implementation function?

**Comment by Trass3r** [comment]
Well the macro approach would be the best in this use-case. If you can make it work somehow then perfect.

PRETTY_FUNCTION is the next best thing for general use (i.e. also works when not using macros in the first place). As long as the compiler supports it it's the best option you got in C++ atm and the basis of https://github.com/Neargye/magic_enum for example.
But it depends on the compiler incl. the version.
I guess you could simply fall back to the rtti based approach on older compilers.

**Comment by cdeln** [comment]
I just tried this out

```cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include <doctest/doctest.h>

struct X {};

TEST_CASE_TEMPLATE("test", T, std::vector<int>, X) {
    CHECK(false);
}
```

which gives the following output

```console
[doctest] doctest version is "2.4.11"
[doctest] run with "--help" for options
===============================================================================
main.cpp:6:
TEST CASE:  test<std::vector<int>>

main.cpp:7: ERROR: CHECK( false ) is NOT correct!
  values: CHECK( false )

===============================================================================
main.cpp:6:
TEST CASE:  test<X>

main.cpp:7: ERROR: CHECK( false ) is NOT correct!
  values: CHECK( false )

===============================================================================
[doctest] test cases: 2 | 0 passed | 2 failed | 0 skipped
[doctest] assertions: 2 | 0 passed | 2 failed |
[doctest] Status: FAILURE!
```

so it seems this works for both built in containers and custom types already without using `TYPE_TO_STRING`.
Does this mean that this issue can be considered complete?

**Comment by onqtam** [comment]
what happens if a typedef is used? I've already forgotten the internals but I think doctest currently does simple stringification of whatever is passed to the macro and that wouldn't resolve properly to an underlying type. I could be wrong.

**Comment by cdeln** [comment]
Yes, that works too. Bisected the source code, this is the PR that fixed it: https://github.com/doctest/doctest/pull/585

The following code
```cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include <doctest/doctest.h>

struct X {};

typedef struct X Y;

TEST_CASE_TEMPLATE("test", T, std::vector<int>, X, Y) {
    CHECK(false);
}
```

gives output

```console
[doctest] doctest version is "2.4.8"
[doctest] run with "--help" for options
===============================================================================
main.cpp:8:
TEST CASE:  test<std::vector<int>>

main.cpp:9: ERROR: CHECK( false ) is NOT correct!
  values: CHECK( false )

===============================================================================
main.cpp:8:
TEST CASE:  test<X>

main.cpp:9: ERROR: CHECK( false ) is NOT correct!
  values: CHECK( false )

===============================================================================
main.cpp:8:
TEST CASE:  test<X>

main.cpp:9: ERROR: CHECK( false ) is NOT correct!
  values: CHECK( false )

===============================================================================
[doctest] test cases: 3 | 0 passed | 3 failed | 0 skipped
[doctest] assertions: 3 | 0 passed | 3 failed |
[doctest] Status: FAILURE!
```

**Comment by cdeln** [comment]
@Saalvage Could you have a final look please?
